PMSCore/UserManager.cs

```diff
@@ -0,0 +1,147 @@
+        /// <param name="logger">Logger instance for logging operations.</param>
+        private UserManager(ILogger? logger)
+        {
+            // Use the provided logger or create a new one if null.
+            _logger = logger ?? new LoggerFactory().CreateLogger("UserManager");
+
+            // Initialize new List to _RegisteredUsers.
+            this._RegisteredUsers = new List<IUser>();
+
+        }
+
+        /// <summary>
+        /// Retrieves the singleton instance of <see cref="UserManager"/>.
+        /// Creates a new instance if it does not already exist.
+        /// </summary>
+        /// <param name="logger">Logger instance for logging operations.</param>
+        /// <returns>The singleton instance of <see cref="UserManager"/>.</returns>
+        public static UserManager GetInstance(ILogger? logger = null)
+        {
+            if (_singleInstance == null)
+            {
+                _singleInstance = new UserManager(logger);
+            }
+
+            return _singleInstance;
+        }
+
+        /// <summary>
+        /// Resets the singleton instance of <see cref="UserManager"/>.
+        /// </summary>
+        public static void ResetInstance()
+        {
+            _singleInstance = null;
+        }
+
+        /// <summary>
+        /// Changes the permission level of a specified user.
+        /// </summary>
+        /// <param name="currentUser">The current system user.</param>
+        /// <param name="targetUser">The user whose permissions are being changed.</param>
+        /// <param name="permission">The new permission level to assign.</param>
+        /// <returns><c>true</c> if the permissions were changed successfully; otherwise, <c>false</c>.</returns>
+        public AssociationStatus ChangeUserPermission(IUser currentUser, IUser targetUser, Permission permission)
+        {
+            // TODO: Class StateHandler to manage a User's permissions access.
+            // StateHandler should validate _currentUser and check for permission level before applying.
+
+            if ((targetUser == null) || (currentUser == null))
+            {
+                _logger.LogError("UserID cannot be null or empty.");
+                return AssociationStatus.InvalidUser;
+            }
+            else if (currentUser.GetPermission() >= Permission.ADMIN || currentUser.Equals(targetUser))
+            {
+                targetUser.SetPermission(permission);
+                return AssociationStatus.NoError;
+            }
+
+            return AssociationStatus.InvalidPermission;
+        }
+
+        /// <summary>
+        /// Logs in the specified user if they are registered.
+        /// </summary>
+        /// <param name="loginUser">The user attempting to log in.</param>
+        /// <returns><see cref="AssociationStatus.NoError"/> if login is successful; otherwise, <see cref="AssociationStatus.InvalidUser"/>.</returns>
+        public AssociationStatus Login(IUser loginUser)
```

Comment:
I agree with this. Do you mind if I wait to implement it until the StateManager is sorted? Will break all of my tests otherwise and I'm not sure how to reroute them to work, only to have to redo it when StateManager is done.